### PR TITLE
core, schedule: use size as a factor of score

### PR DESCRIPTION
**Pull request by Connor1996**
When calculate region/leader score, use size instead of count, in order to balance store size in final.

new shouldBalance logic:

1. introduce avgScore to make sure moving is not unnecessary. Cause in some case:
   note that: avgScore is not equal to totalScore / count, since weight is not same for every store, move will cause the change of score for two involved store is not equal. So it should be totalSize / totalWeight.
   - sourceScore > targetScore > avgScore, if move, targetScore is still higher than avgScore, and then target store will be scheduled to move some region or leader to other stores with smaller score. But is is better to move some region or leader from source store to store with score lower than avgScore, so it is unnecessary
   - avgScore > sourceScore > targetScore, if move, sourceScore is still lower than avgScore, source store still need other stores to move region or leader to it for closing avgScore, so it is unnecessary.
2. avgScore is the score when stores in perfect balance state. Based on the weight, we can know the difference of size for current state to be in perfect balance state, and the minimum of them is the largest size of region to be moved. And the tolerantRatio provides a buffer to make the cluster stable, so that we don't need to schedule very frequently.

### CODE REVIEW

File: server/core/store.go [hunk 1]
```diff
@@ -366,6 +384,19 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
...
+func (s *StoresInfo) AverageResourceScore(kind ResourceKind) float64 {
+	var totalResourceScore float64
+	var count int
+	for _, s := range s.stores {
+		if s.IsUp() {
+			count++
+			totalResourceScore += s.ResourceScore(kind)
+		}
+	}
+	return totalResourceScore / float64(count)
```
Comment (disksing) [inline server/core/store.go]: I think it should be `totalRegionSize / totalStoreWeight`.

File: server/schedulers/utils.go [hunk 2]
```diff
@@ -161,21 +151,26 @@ func takeInfluence(store *core.StoreInfo, storeInfluence *schedule.StoreInfluenc
...
-	return diffCount >= minBalanceDiff(sourceCount)
+
+	// avgScore is the goal for every store
+	// in expection, sourceScore > avgScore > targetScore
+	// if not, moving region is not necessary
+	// In this case, either sourceSizeDiff or targetSizeDiff will be negative, then obviously return false
+	sourceSizeDiff := (sourceScore - avgScore) * source.ResourceWeight(kind)
+	targetSizeDiff := (avgScore - targetScore) * target.ResourceWeight(kind)
+
+	return math.Min(sourceSizeDiff, targetSizeDiff) > float64(region.ApproximateSize)*tolerantRatio
```
Comment (nolouch) [inline server/schedulers/utils.go]: need we dynamic adjust `tolerantRatio` according to the cluster size like before?

File: server/schedulers/utils.go [hunk 3]
```diff
@@ -140,19 +140,9 @@ func minDuration(a, b time.Duration) time.Duration {
 }
 
 const (
-	bootstrapBalanceCount = 10
-	bootstrapBalanceDiff  = 2
+	tolerantRatio = 2.5
```
Comment (disksing) [inline server/schedulers/utils.go]: Maybe add a config variable for it.

**Comment by disksing** [comment]
LGTM.

File: conf/config.toml [hunk 4]
```diff
@@ -51,6 +51,7 @@ max-store-down-time = "1h"
 leader-schedule-limit = 64
 region-schedule-limit = 16
 replica-schedule-limit = 24
+balance-tolerant-ratio = 2.5
```
Comment (siddontang) [inline conf/config.toml]: what does tolerant mean here?

File: server/core/store.go [hunk 5]
```diff
@@ -366,6 +384,19 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
...
+func (s *StoresInfo) AverageResourceScore(kind ResourceKind) float64 {
+	var totalResourceSize int64
+	var totalResourceWeight float64
+	for _, s := range s.stores {
+		if s.IsUp() {
+			totalResourceWeight += s.ResourceWeight(kind)
+			totalResourceSize += s.ResourceSize(kind)
+		}
+	}
+	return float64(totalResourceSize) / totalResourceWeight
```
Comment (siddontang) [inline server/core/store.go]: need to check weight is 0 here?

File: server/schedulers/utils.go [hunk 6]
```diff
@@ -161,21 +147,26 @@ func takeInfluence(store *core.StoreInfo, storeInfluence *schedule.StoreInfluenc
 }
 
 // shouldBalance returns true if we should balance the source and target store.
-// The min balance diff provides a buffer to make the cluster stable, so that we
+// The tolerantRatio provides a buffer to make the cluster stable, so that we
 // don't need to schedule very frequently.
-func shouldBalance(source, target *core.StoreInfo, kind core.ResourceKind, opInfluence schedule.OpInfluence) bool {
+func shouldBalance(source, target *core.StoreInfo, avgScore float64, kind core.ResourceKind, region *core.RegionInfo, opInfluence schedule.OpInfluence, tolerantRatio float64) bool {
```
Comment (siddontang) [inline server/schedulers/utils.go]: too many arguments now

/cc @disksing

File: conf/config.toml (hunk 4 above)
Comment (Connor1996) [reply conf/config.toml]: tolerate size difference

Comment (siddontang) [reply conf/config.toml]: so the better name may be tolerate-size-ratio

File: server/schedulers/utils.go (hunk 6 above)
Comment (siddontang) [reply server/schedulers/utils.go]: Please add a TODO comment here to simplify the arguments here.

File: server/schedule/opts.go [hunk 7]
```diff
@@ -21,6 +21,7 @@ import (
 type Options interface {
 	GetLeaderScheduleLimit() uint64
 	GetRegionScheduleLimit() uint64
+	GetTolerantSizeRatio() float64
```
Comment (siddontang) [inline server/schedule/opts.go]: here you use tolerant size, but your TOML file use tolerate size, please choose one.

**Pull request merged**